### ISSUE TITLE
Feature Request: Enable End-to-End Encryption

### ISSUE THREAD

**Issue by mike-lloyd03** [BlueBubblesApp/bluebubbles-server#421]
All communication is protected by TLS encryption already if BlueBubbles is being served on an https domain. But per our discussion on Discord today, is it possible to enable end-to-end encryption so messages are not readable by Google's push notification servers similar to AirMessage's setup discussed [here](https://airmessage.org/blog/end-to-end-encryption-across-all-of-airmessage)?

**Comment by zlshames** [comment]
Yup this is something that i've discussed doing. It will take both a client update and server update to accomplish this. I will have to agree upon a "standard" format so that the client can properly decrypt the payloads. It needs to be done. I also want to migrate authentication to be oauth-style

**Comment by mike-lloyd03** [comment]
Thanks for being open to the idea. If switching to OAuth, would there still be an option for password based auth? Users like myself would like to keep as many third parties out of the process as possible.

**Comment by zlshames** [comment]
So, if I did move to a more o-auth like workflow, it would still require a "password", however, only for the initial connection. The client would then get a token back that it can use to authenticate with the server. The password will be used for encrypting and decrypting payloads only.

That way, if somehow you get MITM-attacked, the attacker would only have your access token, and not the real password to decrypt the payload. OAuth is theoretically more secure than just password-based auth.

**Comment by mike-lloyd03** [comment]
How much work would it be to enable this? I'd like to contribute but it's been a few years since I've written any JavaScript/TypeScript.

**Comment by mike-lloyd03** [comment]
Most of that seems above my head but I could probably tackle that first point. Adding a new column to the DB and writing the necessary migrations is right in my wheelhouse. Would a PR to get this small step going be okay?

The rest of it would take me getting familiarized with the codebase and reviving my Javascript knowledge.

Also, does this conversation have any connection to #82?

**Comment by zlshames** [comment]
> Most of that seems above my head but I could probably tackle that first point. Adding a new column to the DB and writing the necessary migrations is right in my wheelhouse. Would a PR to get this small step going be okay?
> 
> The rest of it would take me getting familiarized with the codebase and reviving my Javascript knowledge.
> 
> Also, does this conversation have any connection to #82?

It's actually not a new column in the DB. It's just a new entry in the config table in the DB. It's basically just adding one line to the constants file in the database/server/ directory I think. 

That said, it might not be worth making a PR for unless you plan on implementing other code as well. If you're not too familiar with the codebase, I'm not sure this is the best issue to tackle at this point. 

As for #82, it's sorta related. That ticket was done when we used the websocketd for API communication. Now weve moved away from that and don't really use it anymore. But the encryption algorithm was one plus that came from that ticket. Which we can use here

**Comment by mr-ransel** [comment]
Came to circle back on this and see if there's an updated path to using the existing encryption library for Firebase notifications and the regular API. 

I saw that in the v1.8.0 beta release we're entirely disabling encryption automatically over the socket due to a problem with it not working on Android, which made me want to check in and see if we've given more consideration on how to get that functionality migrated to the non-websocket API. I've reviewed the above description and if it's still accurate I'd be interested in collaborating on it to get the server-side ready but updating all the different clients to support it might be out of my depth.

**Comment by zlshames** [comment]
> Came to circle back on this and see if there's an updated path to using the existing encryption library for Firebase notifications and the regular API.
> 
> I saw that in the v1.8.0 beta release we're automatically disabling encryption entirely over the socket due to a problem with it not working on Android, which made me want to check in and see if we've given more consideration on how to get that functionality migrated to the non-websocket API and firebase payloads. I've reviewed the above description and if it's still accurate I'd be interested in collaborating on it to get the server-side ready but updating all the different clients to support it might be out of my depth.

What are firebase's existing encryption library for notifications? Wouldn't it be build into the Firebase SDK by default? Also, I wouldnt want to move completely to Firebase messages purely because some users on web & desktop dont use firebase

**Comment by mr-ransel** [comment]
Dug this up, need to look into the extra examples they cite: 
FCM does not provide an end-to-end solution. However, there are external solutions available such as Capillary or DTLS.  https://firebase.google.com/docs/cloud-messaging/concept-options#data_messages

**Comment by derekmorr** [comment]
Another option is what Signal does - send an empty notification. The notification causes the app to connect to the server and download the message directly.

**Comment by mr-ransel** [comment]
Honestly that's... kind of brilliant tbh. Side steps the entire issue. And if people want to use ngrok or not it's entirely up to them, letting people handle encryption however they want

**Comment by derekmorr** [comment]
I think having one encryption module serve multiple roles would unnecessarily complicate the design. Ideally the payload data would be versioned, and that payload data would be encrypted — providing a separation of concerns.

As to empty vs non-empty notification messages — I don’t really see the benefit of only being able to receive messages if I can’t reply to them. If that were to happened, I would probably just resort to SMS to contact the person, which defeats on the main benefits of iMessage (encrypted comms). 

I’m concerned about Capillary for several reasons. The project looks to be unmaintained, and it has a suspect design. It looks like the client registers a long-term public key with the server, and all messages are encrypted with this key. This is similar to how iMessage and PGP work, which means it lacks Forward Secrecy. While I would much prefer to see an approach which rotates the keys regularly (like the Signal protocol), I’m not sure if it’s worth the extra design complexity since iMessage doesn’t have Forward Secrecy anyway.

**Comment by zlshames** [comment]
> I think having one encryption module serve multiple roles would unnecessarily complicate the design. Ideally the payload data would be versioned, and that payload data would be encrypted — providing a separation of concerns.
> 
> As to empty vs non-empty notification messages — I don’t really see the benefit of only being able to receive messages if I can’t reply to them. If that were to happened, I would probably just resort to SMS to contact the person, which defeats on the main benefits of iMessage (encrypted comms).
> 
> I’m concerned about Capillary for several reasons. The project looks to be unmaintained, and it has a suspect design. It looks like the client registers a long-term public key with the server, and all messages are encrypted with this key. This is similar to how iMessage and PGP work, which means it lacks Forward Secrecy. While I would much prefer to see an approach which rotates the keys regularly (like the Signal protocol), I’m not sure if it’s worth the extra design complexity since iMessage doesn’t have Forward Secrecy anyway.

Our previous encryption would rely on the server password as a "password-based" encryption. So if you want to update the "key", it would be similar to just updating your password. Which you'd need to change on the client side too. I would think we'd do something similar, and not rely on Capillary or anything.

**Comment by cameronaaron** [comment]
I suggest adopting a phased approach:

Token-Based Authentication:

Implement token generation and storage in the server config table.
Create an endpoint (/api/token?type=password) to return the token upon initial authentication.
Update middleware to use tokens instead of passwords.
End-to-End Encryption:

Use the user-set password to encrypt/decrypt API payloads.
Add metadata to payloads to indicate encryption status and method.
Implement empty notifications prompting the client to fetch the actual message from the server, similar to Signal's approach.

**Comment by derekmorr** [comment]
If we do empty notifications, why do we need tokens?
